device/src/main/java/com/softserve/lv460/device/config/ChangeStreamConfig.java

```diff
@@ -0,0 +1,97 @@
+package com.softserve.lv460.device.config;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.softserve.lv460.device.action.Action;
+import com.softserve.lv460.device.action.ActionRegistry;
+import com.softserve.lv460.device.config.cache.RuleCacheConfig;
+import com.softserve.lv460.device.document.DeviceData;
+import com.softserve.lv460.device.dto.rule.ActionRuleDto;
+import com.softserve.lv460.device.dto.rule.RuleDto;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.configurationprocessor.json.JSONException;
+import org.springframework.boot.configurationprocessor.json.JSONObject;
+import org.springframework.context.annotation.Bean;
+import org.springframework.data.mongodb.core.ChangeStreamEvent;
+import org.springframework.data.mongodb.core.ReactiveMongoOperations;
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Flux;
+
+import javax.annotation.PostConstruct;
+import java.util.*;
+import java.util.function.BiPredicate;
+
+@Service
+@AllArgsConstructor
+@Slf4j
+public class ChangeStreamConfig {
+  private final PropertiesConfig propertiesConfig;
+  private final ActionRegistry actionRegistry;
+  private final RuleCacheConfig ruleCacheConfig;
+  private final ReactiveMongoOperations reactiveTemplate;
+
+  @PostConstruct
+  public void subscribeToChangeStream() {
+    Flux<ChangeStreamEvent<DeviceData>> deviceData = reactiveTemplate.changeStream(DeviceData.class)
+            .watchCollection(propertiesConfig.getCollection()).listen();
+
+    deviceData.doOnNext((event) -> {
+      List<RuleDto> rules = ruleCacheConfig.getCache(event.getBody().getUuId());
+      for (RuleDto rule : rules) {
```

Comment:
this can be rewritten with stream().filter( rule -> checkRuleCondition .... ).forEach(this::executeAction)